### PR TITLE
Weaken bounds on Array impls

### PR DESCRIPTION
**Pull request by workingjubilee**
Weaken the bounds on various implementations for Array, to make things easier to refactor. In that vein, clean up a function from on IntoDatum that can be inside Array's IntoDatum impl.

This weakening is mostly irrelevant because you still can't construct such an Array.

### CODE REVIEW

File: pgrx/src/datum/into.rs [hunk 1]
```diff
@@ -34,9 +34,6 @@ pub trait IntoDatum {
     fn composite_type_oid(&self) -> Option<Oid> {
         None
     }
-    fn array_type_oid() -> pg_sys::Oid {
-        unsafe { pg_sys::get_array_type(Self::type_oid()) }
-    }
```
Comment (workingjubilee) [inline pgrx/src/datum/into.rs]: This is technically a breaking change.

**Comment by workingjubilee** [comment]
Actually, I think I'll separate that out.

File: pgrx/src/datum/array.rs [hunk 2]
```diff
@@ -275,6 +243,41 @@ impl<'a, T: FromDatum> Array<'a, T> {
...
+    /// and this can hypothetically cause a memory leak if so.
+    #[inline]
+    pub fn into_array_type(self) -> *const pg_sys::ArrayType {
+        // may be worth replacing this function when Toast<T> matures enough
+        // to be used as a public type with a fn(self) -> Toast<RawArray>
+
+        let Array { raw, .. } = self;
+        // Wrap the Toast<RawArray> to prevent it from deallocating itself
+        let mut raw = core::mem::ManuallyDrop::new(raw);
+        let ptr = raw.deref_mut().deref_mut() as *mut RawArray;
```
Comment (thomcc) [inline pgrx/src/datum/array.rs:259]: you didn't add this, but lol.

**Review by thomcc — APPROVED** [review]

File: pgrx/src/datum/array.rs (hunk 2 above)
Comment (workingjubilee) [reply pgrx/src/datum/array.rs:259]: :see_no_evil:

**Pull request merged**